### PR TITLE
Check if there is a valid session before messing with session ini values

### LINKED ISSUE #12153: E_WARNING in libraries/joomla/session/handler/joomla.php - J3.6.2
Hello guys

In my PHP application, I'm interfacing the Joomla!® framework for retrieving the list of versions and being able to suggest to my users that their version is an old one.  Quite easy.  

By loading the framework; I receive this warning : 
`[E_WARNING] [2016-09-23 13:32:27 (CEST)] [File xxxx/libraries/joomla/session/handler/joomla.php] [Line 45] ==> ini_set(): A session is active. You cannot change the session module's ini settings at this time`

This happens because, in my own code, I've already started a session.

Could be easily solved by checking session_id().
### Steps to reproduce the issue

The current code in /libraries/joomla/session/handler/joomla.php is :

``` php
public function __construct($options = array())
{
   // Disable transparent sid support
   ini_set('session.use_trans_sid', '0');

   // Only allow the session ID to come from cookies and nothing else.
   ini_set('session.use_only_cookies', '1');

   // Set options
   $this->setOptions($options);
   $this->setCookieParams();
}
```
### Expected result

No warning
### Actual result

A warning is displayed : 
`[E_WARNING] [2016-09-23 13:32:27 (CEST)] [File xxxx/libraries/joomla/session/handler/joomla.php] [Line 45] ==> ini_set(): A session is active. You cannot change the session module's ini settings at this time`
### Additional comments

By changing the code like below, the warning dissapear.  Can this change be done in Joomla ?  Don't know if there will be side effect.

``` php
public function __construct($options = array())
{
   if (session_id()==0)
   {
      // Disable transparent sid support
      ini_set('session.use_trans_sid', '0');

      // Only allow the session ID to come from cookies and nothing else.
      ini_set('session.use_only_cookies', '1');
   }

   // Set options
   $this->setOptions($options);
   $this->setCookieParams();
}
```

Thanks a lot.

### LINKED ISSUE #15680: Error:"A session is active. You cannot change the session module's ini settings at this time"
I reopen this issue here because it still exists in 3.7:
Joomla! Issue Tracker | Joomla! CMS #12153 - E_WARNING in libraries/joomla/session/handler/joomla.php - J3.6.2  https://issues.joomla.org/tracker/joomla-cms/12153

### Steps to reproduce the issue
After an update to 3.7

Error message :
"A session is active. You cannot change the session module's ini settings at this time"

Various unsatisfying solutions have been found, like:
comment out: ini_set('session.use_trans_sid', '0');

Could Cavo789's be doable ?

### PR DESCRIPTION
**Pull request by PhilETaylor**
Closes #15680
Closes #12153
Check if there is a valid session before messing with session ini values
(NOT TESTED, coded on sight only - please test!)

### CODE REVIEW

File: libraries/joomla/session/handler/joomla.php [hunk 1]
```diff
@@ -42,11 +42,14 @@ class JSessionHandlerJoomla extends JSessionHandlerNative
 	 */
 	public function __construct($options = array())
 	{
-		// Disable transparent sid support
-		ini_set('session.use_trans_sid', '0');
+		if (session_id() === "")
```
Comment (mbabker) [inline libraries/joomla/session/handler/joomla.php]: If we really have to do this we should be checking either `$this->isStarted()` or `$this->getId()`.  Although I really feel like this is just a edge case workaround and not something that should exist in production code (I'm not aware of any other session API that has this level of checks).

**Comment by PhilETaylor** [comment]
> this is just a edge case workaround and not something that should exist in production code

In that case someone needs to make a decision and then we can either change this PR or close the two outstanding issues with one foul sweep of the brush...

**Comment by PhilETaylor** [comment]
> $this->isStarted()

This will only return true if the session was started by the Joomla API right, and not if "something else" started the session. 

> $this->getId() 

This is nothing more than a wrapper to `session_id()` anyway :)

**Comment by mbabker** [comment]
>This is nothing more than a wrapper to `session_id()` anyway :)

Better to use the abstractions than hardcode stuff 😉 

You have my opinion on the matter, whether others agree or not is to be determined.  Honestly, session packages are one part of the PHP ecosystem that lack interoperability and part of it does boil down to the internal uses of `ini_set()` (at least in the cases of our own Session APIs, Symfony, and Aura; don't have the time at the moment to look for other packages and review those).

**Comment by csthomas** [comment]
IMO this place would be better for such `ini_set()` https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/session/session.php#L132

**Comment by mbabker** [comment]
Not really, the intent with the abstraction layers is to make it so that configuration changes that don't have to be made for an environment aren't.  If you ran a custom app off the CMS app stack using JSession you could use a different `JSessionHandlerInterface` implementation that either doesn't need that `ini_set()` call or makes different ones altogether.

`JSession` (and in 4.0 the Framework's `Joomla\Session\Session`) are really just the interface for folks to read/write data from the session store.  The abstraction layers underneath it (storage and handler) are the internal details where these configs start getting made.  So `JSession` should rightfully be as ignorant to outside influences as practical.

**Comment by PhilETaylor** [comment]
Implemented @mbabker request to use api. 
Implemented @csthomas request to move settings.

**Comment by PhilETaylor** [comment]
Reverted @csthomas request to move settings on @mbabker suggestion

**Comment by csthomas** [comment]
@PhilETaylor I have a new suggestion.

What about move it to `start()` method?

**Comment by PhilETaylor** [comment]
No idea - these kind of decisions are above my pay grade - paging @mbabker :)

**Comment by mbabker** [comment]
That could be viable.

**Comment by PhilETaylor** [comment]
Implemented @csthomas request to move to start() :-)

**Comment by mbabker** [comment]
`JSessionHandlerJoomla::start()` please.  Unless there's some absolute critical reason it HAS to be in `JSession::start()`.

**Comment by PhilETaylor** [comment]
Done.

**Comment by csthomas** [comment]
Great, for me ok

**Comment by csthomas** [comment]
@mbabker I thought about `JSessionHandlerJoomla::start()`, I forgot to mention class name.

**Comment by wetken** [comment]
I have this issue using Joomla 3.7.5. Is there anything I can do to resolve it until the next release?
(I'm using an unconventional setup, I accept that, but I'm also fighting a time constraint.)

Warning: ini_set(): A session is active. You cannot change the session module's ini settings at this time in /home/.../libraries/joomla/session/handler/joomla.php on line 46

Warning: Cannot modify header information - headers already sent by (output started at /home/.../libraries/joomla/session/handler/joomla.php:46) in /home/.../libraries/joomla/session/handler/joomla.php on line 104
Error displaying the error page: Application Instantiation Error: Failed to start the session because headers have already been sent by "/home/.../libraries/joomla/session/handler/joomla.php" at line 46.

**Comment by ghost** [comment]
@wetken please don't post same Comment double like in #[15680](https://github.com/joomla/joomla-cms/issues/15680#issuecomment-328455108)

**Comment by csthomas** [comment]
@wetken It should be easy to copy this file and replace in 3.7.5. This PR was designed for 3.7.
At J3.8 this file was moved to other place.

If you use joomla git on linux and you have shell access you can add this changes by command:
```
curl https://patch-diff.githubusercontent.com/raw/joomla/joomla-cms/pull/15742.diff | git apply
```

**Comment by wetken** [comment]
I’m sorry, Franz, I realised after I had posted on the closed thread, so I re-posted on the other open one, hoping I had done no harm.

If you have a pressing, intractable, issue and you’re not part of the community it can sometimes be difficult to get any attention at all.

Thank you for taking the trouble.

Ken

**Comment by wetken** [comment]
Many thanks, Tomasz, that has helped tremendously.

**Comment by andrewpthompson** [comment]
@PhilETaylor and @mbabker I would like to ask you guys if there is still any plan to implement this PR? It would resolve a problem in the [CiviCRM](https://www.civicrm.org) extension that has been present since Joomla 3.8.0. As CiviCRM has far more Drupal & Wordpress users and developers than it does Joomla, I've been trying to get this fixed out of necessity.

CiviCRM has a cron script that bootstraps the CMS and runs CiviCRM's scheduled jobs. In some environments (PHP 7) it fails to run due to:
```
Warning: ini_set(): A session is active. You cannot change the session module's ini settings at this time in /var/www/html/membership/libraries/joomla/session/handler/joomla.php on line 48
Error: Failed to start application: Failed to start the session because headers have already been sent by "/var/www/html/membership/libraries/joomla/session/handler/joomla.php" at line 48.
```
CiviCRM's [cron.php is doing a session_start()](https://github.com/civicrm/civicrm-core/blob/78a4b4ca77f3e1a95c61b838e925dbb25b7d0541/CRM/Core/Session.php#L127) and this occurs before [it calls JPluginHelper::importPlugin('civicrm');](https://github.com/civicrm/civicrm-core/blob/78a4b4ca77f3e1a95c61b838e925dbb25b7d0541/CRM/Utils/Hook/Joomla.php#L80) which leads to Joomla doing the `ini_set()` in `libraries/joomla/session/handler/joomla.php` when there is a session already active.

Almost the same thing happens with its REST API and again it [does a session_start()](https://github.com/civicrm/civicrm-core/blob/78a4b4ca77f3e1a95c61b838e925dbb25b7d0541/extern/rest.php#L34) before bootstapping Joomla. 

I updated Phil's [PR for Joomla 3.8](https://github.com/andrewpthompson/joomla-cms/commit/f72c0fb4f70181db9f97313f5e5339b3d44cce98) to test this and it does resolve these particular problems (although I don't really know if it's correct for CiviCRM to be doing the session_start() in the first place).

**Comment by PhilETaylor** [comment]
@andrewpthompson If people test it and provide test results then it will have more of a chance of being accepted for merging.

**Comment by andrewpthompson** [comment]
@PhilETaylor Would you be prepared to update your PR to resolve the conflicts so that it's more likely to get tested by other people? Thanks.

**Comment by wetken** [comment]
Sorry, came at me out of the blue, wasn’t sure what was happening for a moment.

What exactly do you want me to do – what’s my PR?

**Comment by andrewpthompson** [comment]
@wetken My question was to PhilETaylor, the author of this pull request.

**Comment by roland-d** [comment]
@PhilETaylor Can you fix the conflicts so I can get it tested over the next two days?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15742">issues.joomla.org/tracker/joomla-cms/15742</a>.</sub>

**Comment by PhilETaylor** [comment]
The joy of having a PR open for over a year. 👎 

It looks like the conflicts come from @mbabker 's commit here https://github.com/joomla/joomla-cms/commit/5b92dc69b39240debc69fb90c7f8d71f060631e8

I will not be able to fix the conflicts this week. This is my busiest week of the year, the first week of the school holidays, and the week everyone wants (paid) work done before I take several weeks 'off' to go on holiday (Tokyo). 

Please feel free to close and replace this PR in my absence.

**Comment by roland-d** [comment]
@PhilETaylor Thank you for replying, I will see what we can do.

**Comment by PhilETaylor** [comment]
https://github.com/joomla/joomla-cms/pull/21260

**Pull request closed without merging**